File: shifai-android/app/src/main/java/com/shifai/domain/intelligence/RuleEngine.kt
package com.shifai.domain.intelligence

import com.shifai.domain.models.*
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlin.math.sqrt

/**
 * Phase 1 Rule Engine — Heuristic-based predictions (Days 1-13)
 * Mirrors iOS RuleEngine.swift
 */
class RuleEngine {

    private val dateFormatter = DateTimeFormatter.ISO_LOCAL_DATE

    /**
     * Predict next period start date using weighted average of last 3 cycles
     */
    fun predictNextPeriod(cycleEntries: List<CycleEntry>): Prediction? {
        val cycleLengths = calculateCycleLengths(cycleEntries)
        if (cycleLengths.size < 2) return null

        val lastCycles = cycleLengths.takeLast(3)
        val weights = if (lastCycles.size == 3) listOf(0.5, 0.3, 0.2) else listOf(0.6, 0.4)
        val weightedAvg = lastCycles.zip(weights).sumOf { (len, w) -> len * w }
        val predictedCycleLength = weightedAvg.roundToInt()

        val stdDev = standardDeviation(cycleLengths.map { it.toDouble() })
        val confidence = (1.0 - (stdDev / 10.0)).coerceIn(0.3, 0.85)

        val lastPeriodStart = findLastPeriodStart(cycleEntries) ?: return null
        val predictedDate = lastPeriodStart.plusDays(predictedCycleLength.toLong())

        return Prediction(
            type = PredictionType.PERIOD_START,
            predictedDate = predictedDate.format(dateFormatter),
            confidence = confidence,
            modelVersion = "rule_engine_v1"
        )
    }

    /**
     * Generate Day 1 Quick Win: sleep benchmark
     */
    fun generateQuickWinDay1(symptoms: List<SymptomLog>): Insight? {
        val sleepLogs = symptoms.filter { it.symptomType == SymptomCategory.SLEEP }
        if (sleepLogs.isEmpty()) return null

        val avgSleep = sleepLogs.map { it.value }.average()
        val benchmarkAvg = 7.0
        val comparison = if (avgSleep >= benchmarkAvg) "meilleur" else "en dessous de"
        val percentage = abs(((avgSleep - benchmarkAvg) / benchmarkAvg) * 100).roundToInt()

        return Insight(
            date = LocalDate.now().format(dateFormatter),
            type = InsightType.QUICK_WIN,
            title = "Ton sommeil est $percentage% $comparison la moyenne 🎉",
            body = "La moyenne recommandée est de ${benchmarkAvg}h. Tu es à ~${"%.1f".format(avgSleep)}h.",
            confidence = 0.9,
            reasoning = "Basé sur tes ${sleepLogs.size} entrées de sommeil.",
            source = IntelligenceSource.RULE_BASED
        )
    }

    /**
     * Generate Day 3 Quick Win: energy mini-pattern
     */
    fun generateQuickWinDay3(symptoms: List<SymptomLog>): Insight? {
        val energyLogs = symptoms
            .filter { it.symptomType == SymptomCategory.ENERGY }
            .sortedBy { it.date }
        if (energyLogs.size < 3) return null

        val trend = energyLogs.last().value - energyLogs.first().value
        val trendText = when {
            trend > 0 -> "augmenté"
            trend < 0 -> "diminué"
            else -> "resté stable"
        }

        return Insight(
            date = LocalDate.now().format(dateFormatter),
            type = InsightType.QUICK_WIN,
            title = "Ton énergie a $trendText ces 3 jours 📈",
            body = "Ton corps suit un rythme. Continue à logger pour découvrir tes patterns !",
            confidence = 0.7,
            reasoning = "Basé sur tes 3 dernières entrées: ${energyLogs.map { it.value }.joinToString(" → ")}",
            source = IntelligenceSource.RULE_BASED
        )
    }

    // ─── Private Helpers ───

    private fun calculateCycleLengths(entries: List<CycleEntry>): List<Int> {
        val periodStarts = entries
            .filter { (it.flowIntensity ?: 0) > 0 }
            .sortedBy { it.date }

        val cycleStarts = mutableListOf<String>()
        var lastDate: String? = null

        for (entry in periodStarts) {
            if (lastDate != null) {
                if (daysBetween(lastDate, entry.date) > 3) {
                    cycleStarts.add(entry.date)
                }
            } else {
                cycleStarts.add(entry.date)
            }
            lastDate = entry.date
        }

        return (1 until cycleStarts.size).map {
            daysBetween(cycleStarts[it - 1], cycleStarts[it])
        }
    }

    private fun findLastPeriodStart(entries: List<CycleEntry>): LocalDate? {
        return entries
            .filter { (it.flowIntensity ?: 0) > 0 }
            .sortedByDescending { it.date }
            .firstOrNull()
            ?.let { LocalDate.parse(it.date, dateFormatter) }
    }

    private fun daysBetween(date1: String, date2: String): Int {
        val d1 = LocalDate.parse(date1, dateFormatter)
        val d2 = LocalDate.parse(date2, dateFormatter)
        return ChronoUnit.DAYS.between(d1, d2).toInt()
    }

    private fun standardDeviation(values: List<Double>): Double {
        val mean = values.average()
        val variance = values.map { (it - mean) * (it - mean) }.average()
        return sqrt(variance)
    }
}

File: shifai-ios/ShifAI/Domain/Intelligence/RuleEngine.swift
import Foundation

// MARK: - Rule Engine (Phase 1 Intelligence)
// Heuristic-based predictions for days 1-13

final class RuleEngine {

    // MARK: - Cycle Prediction

    /// Predict next period start date based on cycle history
    /// Uses weighted average of last 3 cycles (most recent = highest weight)
    func predictNextPeriod(from cycleEntries: [CycleEntry]) -> Prediction? {
        let cycleLengths = calculateCycleLengths(from: cycleEntries)
        guard cycleLengths.count >= 2 else { return nil }

        let lastCycles = Array(cycleLengths.suffix(3))
        let weights: [Double] = lastCycles.count == 3 ? [0.5, 0.3, 0.2] : [0.6, 0.4]
        let weightedAvg = zip(lastCycles, weights).reduce(0.0) { $0 + Double($1.0) * $1.1 }
        let predictedCycleLength = Int(round(weightedAvg))

        // Confidence based on cycle regularity
        let stdDev = standardDeviation(cycleLengths.map { Double($0) })
        let confidence = max(0.3, min(0.85, 1.0 - (stdDev / 10.0)))

        guard let lastPeriodStart = findLastPeriodStart(from: cycleEntries) else { return nil }

        let predictedDate = Calendar.current.date(
            byAdding: .day, value: predictedCycleLength, to: lastPeriodStart
        )

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]

        return Prediction(
            id: UUID().uuidString,
            type: .periodStart,
            predictedDate: predictedDate.map { formatter.string(from: $0) },
            predictedValue: nil,
            confidence: confidence,
            actualDate: nil,
            actualValue: nil,
            accuracyScore: nil,
            modelVersion: "rule_engine_v1",
            createdAt: Date()
        )
    }

    // MARK: - Pattern Detection

    /// Detect simple correlations between tracked data points
    func detectPatterns(symptoms: [SymptomLog], cycles: [CycleEntry]) -> [Insight] {
        var insights: [Insight] = []

        // Pattern: Energy by cycle phase
        if let energyPhaseInsight = detectEnergyPhasePattern(symptoms: symptoms, cycles: cycles) {
            insights.append(energyPhaseInsight)
        }

        // Pattern: Sleep-energy correlation
        if let sleepEnergyInsight = detectSleepEnergyCorrelation(symptoms: symptoms) {
            insights.append(sleepEnergyInsight)
        }

        // Pattern: Stress-pain correlation
        if let stressPainInsight = detectStressPainCorrelation(symptoms: symptoms) {
            insights.append(stressPainInsight)
        }

        return insights
    }

    // MARK: - Quick Wins

    /// Generate Day 1 Quick Win: benchmark sleep vs average
    func generateQuickWinDay1(symptoms: [SymptomLog]) -> Insight? {
        let sleepLogs = symptoms.filter { $0.symptomType == .sleep }
        guard !sleepLogs.isEmpty else { return nil }

        let avgSleep = Double(sleepLogs.map(\.value).reduce(0, +)) / Double(sleepLogs.count)
        let benchmarkAvg = 7.0 // Average for women 25-35

        let comparison = avgSleep >= benchmarkAvg ? "meilleur" : "en dessous de"
        let percentage = abs(Int(((avgSleep - benchmarkAvg) / benchmarkAvg) * 100))

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]

        return Insight(
            id: UUID().uuidString,
            date: formatter.string(from: Date()),
            type: .quickWin,
            title: "Ton sommeil est \(percentage)% \(comparison) la moyenne 🎉",
            body: "La moyenne de sommeil recommandée est de \(benchmarkAvg)h. Tu es à ~\(String(format: "%.1f", avgSleep))h.",
            confidence: 0.9,
            reasoning: "Basé sur tes \(sleepLogs.count) entrées de sommeil, comparé aux recommandations pour ta tranche d'âge.",
            source: .ruleBased
        )
    }

    /// Generate Day 3 Quick Win: mini-pattern detection
    func generateQuickWinDay3(symptoms: [SymptomLog]) -> Insight? {
        let energyLogs = symptoms.filter { $0.symptomType == .energy }.sorted { $0.date < $1.date }
        guard energyLogs.count >= 3 else { return nil }

        let trend = energyLogs.last!.value - energyLogs.first!.value
        let trendText = trend > 0 ? "augmenté" : (trend < 0 ? "diminué" : "resté stable")

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]

        return Insight(
            id: UUID().uuidString,
            date: formatter.string(from: Date()),
            type: .quickWin,
            title: "Ton énergie a \(trendText) ces 3 jours 📈",
            body: "Ton corps suit un rythme. Continue à logger pour découvrir tes patterns !",
            confidence: 0.7,
            reasoning: "Basé sur tes 3 dernières entrées d'énergie: \(energyLogs.map { String($0.value) }.joined(separator: " → "))",
            source: .ruleBased
        )
    }

    // MARK: - Private Helpers

    private func calculateCycleLengths(from entries: [CycleEntry]) -> [Int] {
        let periodStarts = entries
            .filter { ($0.flowIntensity ?? 0) > 0 }
            .sorted { $0.date < $1.date }

        // Group consecutive period days into cycle starts
        var cycleStarts: [String] = []
        var lastDate: String?

        for entry in periodStarts {
            if let last = lastDate {
                // If more than 3 days gap, it's a new cycle
                if daysBetween(last, entry.date) > 3 {
                    cycleStarts.append(entry.date)
                }
            } else {
                cycleStarts.append(entry.date)
            }
            lastDate = entry.date
        }

        // Calculate lengths between consecutive starts
        var lengths: [Int] = []
        for i in 1..<cycleStarts.count {
            if let days = daysBetween(cycleStarts[i-1], cycleStarts[i]) as Int? {
                lengths.append(days)
            }
        }
        return lengths
    }

    private func findLastPeriodStart(from entries: [CycleEntry]) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]

        return entries
            .filter { ($0.flowIntensity ?? 0) > 0 }
            .sorted { $0.date > $1.date }
            .first
            .flatMap { formatter.date(from: $0.date) }
    }

    private func daysBetween(_ dateStr1: String, _ dateStr2: String) -> Int {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        guard let d1 = formatter.date(from: dateStr1),
              let d2 = formatter.date(from: dateStr2) else { return 0 }
        return Calendar.current.dateComponents([.day], from: d1, to: d2).day ?? 0
    }

    private func standardDeviation(_ values: [Double]) -> Double {
        let mean = values.reduce(0, +) / Double(values.count)
        let variance = values.map { pow($0 - mean, 2) }.reduce(0, +) / Double(values.count)
        return sqrt(variance)
    }

    private func detectEnergyPhasePattern(symptoms: [SymptomLog], cycles: [CycleEntry]) -> Insight? {
        // TODO: Correlate energy values with cycle phases
        return nil
    }

    private func detectSleepEnergyCorrelation(symptoms: [SymptomLog]) -> Insight? {
        // TODO: Check if high sleep → high energy next day
        return nil
    }

    private func detectStressPainCorrelation(symptoms: [SymptomLog]) -> Insight? {
        // TODO: Check if high stress → high pain
        return nil
    }
}
